analyzers/src/SonarAnalyzer.CSharp/Rules/UseCachedRegex.cs

```diff
@@ -0,0 +1,194 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class UseCachedRegex : SonarDiagnosticAnalyzer<SyntaxKind>
+{
+    private const string DiagnosticId = "S6614";
+
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+    protected override string MessageFormat => $"{nameof(Regex)} instances should be cached";
+
+    private static readonly HashSet<SyntaxKind> CorrectContextSyntaxKinds = new()
+    {
+        SyntaxKind.MethodDeclaration,
+        SyntaxKind.ConstructorDeclaration,
+        SyntaxKind.DestructorDeclaration,
+        SyntaxKind.GetAccessorDeclaration,
+        SyntaxKind.SetAccessorDeclaration,
+        SyntaxKindEx.LocalFunctionStatement,
+        SyntaxKind.SimpleLambdaExpression,
+        SyntaxKind.ParenthesizedLambdaExpression,
+        SyntaxKind.AnonymousMethodExpression
+    };
+
+    public UseCachedRegex() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var objectCreation = c.Node as ObjectCreationExpressionSyntax;
+                ImplicitObjectCreationExpressionSyntaxWrapper? implicitObjectCreation = objectCreation is null ? (ImplicitObjectCreationExpressionSyntaxWrapper)c.Node : null;
+
+                if (IsCorrectName(c.Node)
+                    && IsWithinCorrectContext(c.Node, out var kindContext)
+                    && IsCorrectObjectType(c.Node, c.SemanticModel)
+                    && IsArgumentConstantOrReadOnly(objectCreation?.ArgumentList ?? implicitObjectCreation?.ArgumentList, c.SemanticModel)
+                    && !IsCompliantAssignment(c.Node, c.SemanticModel, kindContext))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                }
+            },
+            SyntaxKind.ObjectCreationExpression,
+            SyntaxKindEx.ImplicitObjectCreationExpression);
+
+    private static bool IsCorrectName(SyntaxNode node) =>
+        node.NameIs(nameof(Regex)) || node is { RawKind: (int)SyntaxKindEx.ImplicitObjectCreationExpression }; // We check the TypeInfo in IsCorrectObjectType
```

Comment:
I don't really like checking the type info in a method that is supposed to check only for the name. WDYT? Maybe go through the references of other rules using Language.SyntaxKind.ObjectCreationExpressions to see if and how they are doing similar syntax checks and get inspired